### ISSUE TITLE
My regression model is predicting the means of each dimension

### ISSUE THREAD

**Issue by keunwoochoi** [fchollet/keras#1462]
I have a convnet based regression model, as below:

```
model = Sequential()
for i in range(5):
    if i ==0:
        model.add(Convolution2D(input_shape=(1, height, width))) # with or without l2 regularisation 
    else:
        model.add(Convolution2D()) # with or without l2 regularisation 
    model.add(BatchNormalization())
    model.add(PReLU())
    model.add(MaxPooling2D()) # it has 4-6 conv layers.
# then,
model.add(Flatten())
# now fully-connected layers
for j in range(2):
    model.add(Dense(128)) # with or without l2 regularisation
    model.add(BatchNormalization())
    model.add(PReLU()) # 
# then,
model.add(Dense(output_dimension, activation='sigmoid')) # target values is continuous in [0,1]
# then compile with mse and optimisers 
```

The system quite quickly converge to predict just the mean of each dimension as if there's no correlation at all with input data and target vectors, which might be true but I doubt it. I feel like I designed a very inefficient average computation network.

I spent quite a long time inspecting if I made some errors and now really, really frustrated. This is what I have seen.
- After initialisation and before training, I ran `model.predict(test_x)`, which thrown out [0.5]*output_dimensions e.g. [0.5, 0.5, 0.5, 0.5, 0.5] for the all data points. Is it something expected? I initialised all the layers convolution and dense layers with `he_normal`. As I used `sigmoid` as an activation of output layer, the value of 0.5 means it's predicting zeros for all dimension before its activation. I don't have any idea how it could happen with random initialisation - probably it's because of BN? Or is it something that shouldn't happen? I'm so confused.
- By visualising kernels I could see them evolve but the change gets smaller as it backprops. i.e. change in the first layer seems very small, although number says it's being updated. 
- `model.layers[idx].get_weights()` returns a list. Is it [conv_layer_weights, bias]? By its dimension it must be, but just to confirm. If so, is the bias initialised with 0?
- After converging, as the title says, it predicts approximated means of each dimension like this:

```
array([[ 0.40856639,  0.41575322,  0.33291921,  0.06373377,  0.18987735],
       [ 0.40825802,  0.41566607,  0.33345419,  0.06351992,  0.18989056],
       [ 0.40959656,  0.41564348,  0.33396491,  0.06293178,  0.19023168],
       ...,
       [ 0.40923968,  0.41534185,  0.33385843,  0.06460324,  0.19168563],
       [ 0.40915081,  0.41577771,  0.33324537,  0.06356006,  0.18983789],
       [ 0.40562034,  0.41507864,  0.33129829,  0.0580974 ,  0.18172719]])
```
- some other info)
- I'm using Keras 0.3.0 on GPU. Number of data is around 23k, spectrograms of music. + also tested on 0.3.1
- I measured the rates of change of weights. They seems in reasonable range and more importantly, every layers are being updated, so the gradient is flowing okay. (With ReLU however, it suffers from gradient vanishing.)
- I tried to change it to predict only one dimensional value, still it worked the same.
- Different loss functions result in the same predictions.

**Comment by Js-Mim** [comment]
I guess that the batch size you are using is quite big for regressing through short-time Fourier transform frames.

**Comment by keunwoochoi** [comment]
Hi, I haven't thought about this as a batch size issue. Could you let me know how it would be related with batch size, audio, and regression? I was putting 16-48 spectrograms into a batch.

**Comment by Js-Mim** [comment]
By using mini-batches you are actually making the updates on a portion of your data, which contains quite some time variation. In the sense of mini-batch gradient descent, usually the parameters are taken over your samples(not STFT sub-bands/bins) summed and divided by your desired size of the batch (see : https://class.coursera.org/ml-003/lecture/106).
I have also encountered such issues recently and tackled them by decreasing this parameter.
Good rule of thumb : batch size <= overlapped STFT frames

Please keep me posted.

**Comment by keunwoochoi** [comment]
I see. But this is not a frame-based DNN, it's a spectrogram-based CNN, which starts from many spectrogram in a mini batch i.e. (#num_data, 1, #freq_band, #time_frame) in a batch, where spectrograms are individually labeled.
But still I'm not sure I'm fully utilising each data point with my batch size. Will try with smaller batch size and post here. Thanks.

**Comment by Js-Mim** [comment]
Oh i see, now your structure for the CNN is clear to me. 
Not much experience with CNNs and audio data, sorry!
It would be great to post your findings & workarounds.
Cheers,
S.

**Comment by keunwoochoi** [comment]
No worries, actually I keep thinking of it and it could be the reason, I should look into it. I'll keep it posted here.

**Comment by mininaNik** [comment]
Hi @keunwoochoi 
I have similar problem (but in lstm). Have you found the reason of your issue?

Regards,
Mina

**Comment by keunwoochoi** [comment]
Hi @mininaNik, I changed my dataset to the other one - which is supposed to be similar to the original one - and then it started to work.
I don't think small changes (or even mistakes) on hyper-parameters shouldn't affect the network to behave in that way. Even with a very inappropriate setting (i.e. inefficient metrics or loss functions or a model with tiny capacity) didn't show such a result with this new dataset. I still have to figure it out but I would like to suggest you to check out if the data and labels have any correlation that can be learned. I suspect I made mistake on generating the dataset.
